lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+  unsigned int num_args =
+      (unsigned int)(function.args().end() - function.args().begin());
+  for (unsigned int i = 1; i <= num_args; i++) {
+    if (!param_names.count(i)) {
+      param_names[i] = "param" + std::to_string(i);
+    }
+  }
+
+  return param_names;
+}
+
+// TODO: This doesn't really fit here so I will need to find a better place for
+// this. Translates an llvm::Type to a type that conforms to the spec in
+// TypeParser.cpp
+std::string TranslateType(const llvm::Type &type) {
+  unsigned int id = type.getTypeID();
+
+  std::string ret;
+
+  // Types should match the type parsing in TypeParser.cpp
+  switch (id) {
+    case llvm::Type::VoidTyID: {
+      ret = "v";
+      break;
+    }
+    case llvm::Type::HalfTyID: {
+      ret = "e";
+      break;
+    }
+    case llvm::Type::FloatTyID: {
+      ret = "f";
+      break;
+    }
+    case llvm::Type::DoubleTyID: {
+      ret = "d";
+      break;
+    }
+    case llvm::Type::X86_FP80TyID: {
+      ret = "D";
+      break;
+    }
+    case llvm::Type::X86_MMXTyID: {
+      ret = "M";
+      break;
+    }
+    case llvm::Type::IntegerTyID: {
+      auto derived = llvm::cast<llvm::IntegerType>(type);
+      // Since there is no way to check for uint vs. int, lower all integer
+      // types to signed integers. Maybe come back when there is a better way to
+      // figure out unsigned vs signed ints
+      auto sign = true;
+      switch (derived.getBitWidth()) {
+        case 8: {
+          ret = sign ? "b" : "B";
+          break;
+        }
+        case 16: {
+          ret = sign ? "h" : "H";
+          break;
+        }
+        case 32: {
+          ret = sign ? "i" : "I";
+          break;
+        }
+        case 64: {
+          ret = sign ? "l" : "L";
+          break;
+        }
+      }
+      break;
+    }
+    case llvm::Type::FunctionTyID: {
+      ret = "func";
+      break;
+    }
+    case llvm::Type::StructTyID: {
+      auto struct_ptr = llvm::cast<llvm::StructType>(&type);
+      std::string element_list = "";
+      for (unsigned i = 0; i < struct_ptr->getNumElements(); i++) {
+        // TODO: BUG: this if statement protects against infinite chained
+        // pointers, which this code has a hard time with. Notably, this
+        // includes some types of C++ ostream functions.
+        if (struct_ptr->getElementType(i)->isPtrOrPtrVectorTy()) {
+          element_list += "?";
+          continue;
+        }
+        element_list += TranslateType(*struct_ptr->getElementType(i));
+      }
+      ret += " {" + element_list + "}";
+
+      break;
+    }
+    case llvm::Type::ArrayTyID: {
+      ret = "array";
+      break;
+    }
+    case llvm::Type::PointerTyID: {
+      ret = "*";
+      auto derived = llvm::dyn_cast<llvm::PointerType>(&type);
+      // Get the type of the pointee
+      ret += TranslateType(*derived->getElementType());
+      break;
+    }
+
+    default:
+      LOG(ERROR) << "Could not translate TypeID: " << id;
```

Comment:
We can do a better job here. I recommend passing down a data layout, then you can ask the data layout for the type storage size, and then fill in an array of bytes type here.

Reply:
I made changes in 0bbd0d9 that should help the issue. Basically I have two interfaces, one that takes a Datalayout and one that doesn't. I didn't really make sense to me to just have one that accepts a Datalayout because then it would be difficult to serialize a ValueDecl in isolation. So if there is a Datalayout, we translate an unknown type into an array of reasonable size, otherwise, we fail out.

Reply:
Thinking over more, I think `DataLayout` should be a requirement, and we can "force" it using `const llvm::DataLayout &` as a parameter type. I like reference parameters when you want to communicate "never null" for a pointer.

Reply:
Fixed 327d728